File: src/hivClass.cpp
#include "Classes.hpp"

void hivC::aging(const boost2D& ag_prob, Views& v, const StateSpace& s) {
  for (int i = 0; i < N; i++)
    *(at_this + i) = *(at_prev + i);
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++)
        if (s.MODEL == 2 && agr < s.hDB)
          data_db[s.year][sex][agr][cd4] = data_db[s.year-1][sex][agr][cd4];
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG - 1; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        double nHup = v.pre_hiv[sex][agr][cd4] * ag_prob[sex][agr];
        v.now_hiv[sex][agr][cd4]   -= nHup;
        v.now_hiv[sex][agr+1][cd4] += nHup;
        if (s.MODEL == 2 && agr < s.hDB - 1) {
          nHup = data_db[s.year-1][sex][agr][cd4] * ag_prob[sex][agr];
          data_db[s.year][sex][agr][cd4]   -= nHup;
          data_db[s.year][sex][agr+1][cd4] += nHup;
        }
      }
}

void hivC::add_entrants(const dvec& artYesNo, Views& v,
                        const Parameters& p, const StateSpace& s) {
  for (int sex = 0; sex < s.NG; sex++)
    for (int cd4 = 0; cd4 < s.hDS; cd4++) {
      double add = p.ph.paedsurv_cd4dist[s.year][sex][cd4] * artYesNo[sex+2];
      if (s.MODEL == 1)
        v.now_hiv[sex][0][cd4] += add;
      if (s.MODEL == 2) // add to virgin then debut
        data_db[s.year][sex][0][cd4] += add;
    }
}

void hivC::sexual_debut(Views& v, const Parameters& p, const StateSpace& s) {
  for (int sex = 0; sex < s.NG; sex++)
    for (int adb = 0; adb < s.hDB; adb++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        double n_db = data_db[s.year][sex][adb][cd4] * p.ic.db_rate[s.year][sex][adb];
        v.now_hiv[sex][adb][cd4]       += n_db;
        data_db[s.year][sex][adb][cd4] -= n_db;
      }
}

void hivC::deaths (const boost2D& survival_pr, Views& v, const StateSpace& s) {
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        v.now_hiv[sex][agr][cd4] *= survival_pr[sex][agr];
        if (s.MODEL == 2 && agr < s.hDB)
          data_db[s.year][sex][agr][cd4] *= survival_pr[sex][agr];
      }
}

void hivC::migration (const boost2D& migration_pr, Views& v, const StateSpace& s) {
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        v.now_hiv[sex][agr][cd4] *= migration_pr[sex][agr];
        if (s.MODEL == 2 && agr < s.hDB)
          data_db[s.year][sex][agr][cd4] *= migration_pr[sex][agr];
      }
}

void hivC::update_infection (const boost2D& new_infect,
                             const Parameters& p, const StateSpace& s) {
  zeroing(grad); // reset every time step
  zeroing(infect_by_agrp_);
  for (int sex = 0; sex < s.NG; ++sex) {
    int current_age_group = s.ag_[0]; // first age group
    for (int age = 0; age < s.pAG; ++age) {
      if ( s.ag_[age] != current_age_group)
        ++current_age_group;
      infect_by_agrp_[sex][current_age_group-1] += new_infect[sex][age];
    } // end age-groups
  }
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++)
        grad[sex][agr][cd4] += 
          p.nh.cd4_initdist[sex][agr][cd4] * infect_by_agrp_[sex][agr];
}

void hivC::scale_cd4_mort(artC& artpop, Views& v,
                           const Parameters& p, const StateSpace& s) {
  double num, den = 0;
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        num = v.now_hiv[sex][agr][cd4];
        if (s.MODEL==2)
          num += data_db[s.year][sex][agr][cd4];
        for (int dur = 0; dur < s.hTS; dur++) {
          den += v.now_art[sex][agr][cd4][dur];
          if (s.MODEL==2) 
            den += artpop.data_db[s.year][sex][agr][cd4][dur];
        }
        num = ((num + den) == 0.0) ? 1.0 : num / (num + den);
        cd4_mort_[sex][agr][cd4] = num * p.nh.cd4_mort[sex][agr][cd4];
        den = 0.0;
      }
}

void hivC::grad_progress(Views& v, const Parameters& p, const StateSpace& s) {
  if (p.ic.eppmod == 2)
    zeroing(grad); // reset every time step
  if (s.MODEL == 2)
    zeroing(grad_db); // reset, this's the 1st time grad_db is used
  // remove cd4 stage progression (untreated)
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++) {
      for (int cd4 = 0; cd4 < s.hDS - 1; cd4++) {
        double nHup = v.now_hiv[sex][agr][cd4] * p.nh.cd4_prog[sex][agr][cd4];
        death_[sex][agr][cd4]  =
          v.now_hiv[sex][agr][cd4] * cd4_mort_[sex][agr][cd4];
        grad[sex][agr][cd4]   -= (nHup + death_[sex][agr][cd4]);
        grad[sex][agr][cd4+1] += nHup;
        if (s.MODEL == 2 && agr < s.hDB) {
          nHup = data_db[s.year][sex][agr][cd4] * p.nh.cd4_prog[sex][agr][cd4];
          death_db_[sex][agr][cd4] =
            data_db[s.year][sex][agr][cd4] * cd4_mort_[sex][agr][cd4];
          grad_db[sex][agr][cd4]   -= (nHup + death_db_[sex][agr][cd4]);
          grad_db[sex][agr][cd4+1] += nHup;
        }
      }
      death_[sex][agr][s.hDS-1] = 
        v.now_hiv[sex][agr][s.hDS-1] * cd4_mort_[sex][agr][s.hDS-1];
      grad[sex][agr][s.hDS-1] -= death_[sex][agr][s.hDS-1];
      if (s.MODEL == 2 && agr < s.hDB) {
        death_db_[sex][agr][s.hDS-1] =
          data_db[s.year][sex][agr][s.hDS-1] * cd4_mort_[sex][agr][s.hDS-1];
        grad_db[sex][agr][s.hDS-1] -= death_db_[sex][agr][s.hDS-1];
      }
    }
}

void hivC::distribute_artinit (boost3D& artinit, artC& artpop,
                               Views& v, const StateSpace& s) {
    double debut_now, all_hivpop, pr_weight_db, n_artinit_db;
    boost3D artinit_db(extents[s.NG][s.hAG][s.hDS]);
    for (int sex = 0; sex < s.NG; sex++)
      for (int agr = 0; agr < s.hAG; agr++)
        for (int cd4 = 0; cd4 < s.hDS; cd4++) {
          debut_now =
            data_db[s.year][sex][agr][cd4] + s.DT * grad_db[sex][agr][cd4];
          all_hivpop =
            (v.now_hiv[sex][agr][cd4] + s.DT * grad[sex][agr][cd4]) + debut_now;
          if (artinit[sex][agr][cd4] > all_hivpop)
            artinit[sex][agr][cd4]   = all_hivpop;
          pr_weight_db               = debut_now / all_hivpop;
          n_artinit_db               = artinit[sex][agr][cd4] * pr_weight_db;
          artinit_db[sex][agr][cd4]  = n_artinit_db;
          artinit[sex][agr][cd4]    -= n_artinit_db;
          grad_db[sex][agr][cd4]    -= n_artinit_db / s.DT;
        }
    artpop.grad_db_init(artinit_db, s);
}

void hivC::add_grad_to_pop (Views& v, const StateSpace& s) {
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        v.now_hiv[sex][agr][cd4] += s.DT * grad[sex][agr][cd4];
        if (s.MODEL == 2 && agr < s.hDB)
          data_db[s.year][sex][agr][cd4] += s.DT * grad_db[sex][agr][cd4];
      }
}

void hivC::adjust_pop(const boost2D& adj_prob, Views& v, const StateSpace& s) {
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        v.now_hiv[sex][agr][cd4] *= adj_prob[sex][agr];
        if (s.MODEL == 2 && agr < s.hDB)
          data_db[s.year][sex][agr][cd4] *= adj_prob[sex][agr];
      }
}

boost2D hivC::n_by_agr(Views& v, const Parameters& p, const StateSpace& s) {
  boost2D out(extents[s.NG][s.hAG]); 
  for (int sex = 0; sex < s.NG; sex++)
    for (int agr = 0; agr < s.hAG; agr++)
      for (int cd4 = 0; cd4 < s.hDS; cd4++) {
        out[sex][agr] += v.now_hiv[sex][agr][cd4];
        if (s.MODEL == 2)
          out[sex][agr] += data_db[s.year][sex][agr][cd4];
      }
  return out;
}

File: R/hivClass.R
# HIV CLASS
# -----------------------------------------------------------------------------
hivEPP <- R6::R6Class("hivepp", class=F, cloneable=F, portable=F, inherit=eppFP,
    lock_objects=F, 
    public = list(
        year = 1,
        MODEL = "integer",
        data = "array",
        data_db = "array",
        grad = "array",
        grad_db = "array",
        f_death = "array",
        f_death_db = "array",
        initialize = function(fp, MODEL) {
            super$initialize(fp)
            MODEL <<- MODEL
            data <<- array(0, c(hDS, hAG, NG, PROJ_YEARS))
            f_death <<- grad <<- array(0, c(hDS, hAG, NG))
            if (MODEL==2) {
                data_db <<- data
                grad_db <<- grad
                f_death_db <<- f_death
            }
        })
)

hivFunc <- c(
aging = function(ag_prob) {
    data[,,,year]     <<- data[,,,year-1]
    nHup <- sweep(data[,-hAG,, year], 2:3, ag_prob[-hAG,], "*")
    data[,-hAG,,year] <<- data[,-hAG,,year] - nHup
    data[,  -1,,year] <<- data[,  -1,,year] + nHup
    if (MODEL==2) {
        data_db[,,,year]     <<- data_db[,,,year-1]
        nHup <- sweep(data_db[,-hAG,, year], 2:3, ag_prob[-hAG,], "*")
        data_db[,-hAG,,year] <<- data_db[,-hAG,,year] - nHup
        data_db[,  -1,,year] <<- data_db[,  -1,,year] + nHup
    }
},

add_entrants = function(artYesNo) {
    artNO = tail(artYesNo,2)
    if (MODEL==1)
        data[,1,,year] <<- data[,1,,year] + 
            sweep(p$paedsurv_cd4dist[,,year], 2, artNO, "*")
    if (MODEL==2) # add to virgin then debut
        data_db[,1,,year] <<- data_db[,1,,year] + 
            sweep(p$paedsurv_cd4dist[,,year], 2, artNO, "*")
},

sexual_debut = function() {
    debut_hiv <- sweep(data_db[,db_agr,, year], 2:3, p$db_pr, '*')
       data[,db_agr,,year] <<-    data[,db_agr,,year] + debut_hiv
    data_db[,db_agr,,year] <<- data_db[,db_agr,,year] - debut_hiv
},

deaths = function(survival_pr) {
    data[,,,year] <<- sweep(data[,,,year], 2:3, survival_pr, "*")
    if (MODEL==2)
        data_db[,,,year] <<- sweep(data_db[,,,year], 2:3, survival_pr, "*")
},

migration = function(migration_pr) {
    data[,,,year] <<- sweep(data[,,,year], 2:3, migration_pr, "*")
    if (MODEL==2)
        data_db[,,,year] <<- sweep(data_db[,,,year], 2:3, migration_pr, "*")
},

update_infection = function(new_infect) {
    grad[,,] <<- 0 # reset every time step
    grad <<- grad + sweep(p$cd4_initdist, 2:3, sumByAGs(new_infect, ag.idx), "*")
},

grad_progress = function(mortality_rate) { # HIV gradient progress
    if (p$eppmod == "directincid")
        grad[,,] <<- 0 # reset every time step
    # remove cd4 stage progression (untreated)
    nARTup <- p$cd4_prog * data[-hDS,,,year]
    grad[-hDS,,] <<- grad[-hDS,,] - nARTup
    grad[-1,,]   <<- grad[-1,,]   + nARTup # add 
    f_death <<- mortality_rate * data[,,,year]
    grad <<- grad - f_death # HIV mortality, untreated
    if (MODEL==2) {
        grad_db[,,] <<- 0 # reset every time, this's the 1st time grad_db is used
        nARTup <- p$cd4_prog * data_db[-hDS,,,year]
        grad_db[-hDS,,] <<- grad_db[-hDS,,] - nARTup
        grad_db[-1,,]   <<- grad_db[-1,,]   + nARTup 
        f_death_db <<- mortality_rate * data_db[,,,year]
        grad_db <<- grad_db - f_death_db
    }
},

eligible_for_art = function() {
    1 - (1 - rep(0:1, times=c(      p$artcd4elig_idx[year]-1,
                              hDS - p$artcd4elig_idx[year]+1))) * 
        (1 - rep(c(0, p$who34percelig),
                 c(2, hDS-2))) * 
        (1 - rep(p$specpop_percelig[year], hDS))
},

distribute_artinit = function(artinit, artpop) {
    debut_now     <- data_db[,,,year] + DT * grad_db
    all_hivpop    <- (data[,,,year] + DT * grad) + debut_now
    artinit       <- pmin(artinit, all_hivpop)
    pr.weight_db  <- debut_now / all_hivpop
    artinit_db    <- artinit * pr.weight_db
    artinit       <- artinit - artinit_db
    grad_db      <<- grad_db - artinit_db / DT
    artpop$grad_db_init(artinit_db)
    artinit
},

add_grad_to_pop = function() {
    data[,,,year] <<- data[,,,year] + DT * grad
    if (MODEL==2)
        data_db[,,,year] <<- data_db[,,,year] + DT * grad_db
},

adjust_pop = function(adj_prob) {
    data[,,,year] <<- sweep(data[,,,year], 2:3, adj_prob, "*")
    if (MODEL==2)
        data_db[,,,year] <<- sweep(data_db[,,,year], 2:3, adj_prob, "*")
},

# avoid default set method conflict
set_data = function(FUN="+", x, DS=T, AG=T, NG=T, YEAR=NULL) {
    FUN <- match.fun(FUN)
    if (is.null(YEAR))
        data[DS, AG, NG] <<- FUN(data[DS, AG, NG], x)
    else 
        data[DS, AG, NG, YEAR] <<- FUN(data[DS, AG, NG, YEAR], x)
}, 

get = function(YEAR=NULL, DS=T, AG=T, NG=T) {
    'get does not change anything, just return the data, use set to change'
    if (is.null(YEAR))
        data[DS, AG, NG, TRUE] # get all years
    else 
        data[DS, AG, NG, YEAR]
}, 

sweep_sex = function(FUN="*", x, year) {
    data[,,,year] <<- sweep(data[,,,year], 2:3, x, FUN)
})

setMembers(hivEPP, "public", names(hivFunc), hivFunc)